src/multi.c

```diff
@@ -379,14 +402,27 @@ void touchAllWatchedKeysInDb(redisDb *emptied, redisDb *replaced_with) {
     dictIterator *di = dictGetSafeIterator(emptied->watched_keys);
     while((de = dictNext(di)) != NULL) {
         robj *key = dictGetKey(de);
-        if (dictFind(emptied->dict, key->ptr) ||
+        int exists_in_emptied = dictFind(emptied->dict, key->ptr) != NULL;
+        if (exists_in_emptied ||
             (replaced_with && dictFind(replaced_with->dict, key->ptr)))
         {
             list *clients = dictGetVal(de);
             if (!clients) continue;
             listRewind(clients,&li);
             while((ln = listNext(&li))) {
-                client *c = listNodeValue(ln);
+                watchedKey *wk = listNodeValue(ln);
+                if (wk->expired &&
+                    (!exists_in_emptied ||
+                     keyIsExpired(emptied, key)) &&
+                    (!replaced_with ||
+                     !dictFind(replaced_with->dict, key->ptr) ||
+                     keyIsExpired(replaced_with, key)))
+                {
+                    /* The watched key was expired when WATCH was called and now
+                     * it's expired or deleted, i.e. no logical change. */
+                    continue;
+                }
+                client *c = wk->client;
```

Comment:
I considered it, but I think it depends on why we're here. It may be that the key is still stale. Then we shouldn't clear it. If it's deleted, then it doesn't matter if the flag is set or not.